### ISSUE TITLE
[Feature][Op] GQA kernel variants: RoPE, ALiBi, FP8, LogitsClip, SplitKV

### ISSUE THREAD

**Issue by superAngGao** [tile-ai/TileOPs#719]
## Sub-issues

| # | Issue | Scope | Depends on | Priority |
|---|-------|-------|-----------|----------|
| 1 | #722 | SplitKV refactor (decode kernels) | — | P0 (prerequisite) |
| 2 | #723 | RoPE kernels + Op dispatch | #722 | P0 |
| 3 | #724 | LogitsClip kernels | #723 | P2 |
| 4 | #725 | Sliding window file reorganization | — | Independent |
| 5 | #726 | FP8 + RoPE kernels | #723 | P1 |
| 6 | #727 | ALiBi kernels | — | P4 |
| 7 | #720 | FlashInfer benchmark baselines | #723+ | Parallel |

### Dependency graph

```
#725 (SW reorg)           independent
#722 (SplitKV) → #723 (RoPE) → #724 (LogitsClip)
                    │
                    ├──→ #726 (FP8+RoPE)
                    └──→ #720 (FlashInfer baselines)
#727 (ALiBi)              independent
```

---

## Summary

Add comprehensive GQA kernel variants to support RoPE, ALiBi, FP8 quantization, logits soft capping, and explicit split-KV, covering the full combination matrix needed by production LLMs.

## Motivation

Current GQA kernels only support basic attention (no positional encoding, no quantization, no logits clipping). All major LLMs require at least one of these features:

| Model | RoPE | ALiBi | FP8 | LogitsClip | Sliding Window |
|-------|:----:|:-----:|:---:|:----------:|:--------------:|
| LLaMA 3 | ✓ | | ✓ (inference) | | |
| Mistral/Mixtral | ✓ | | ✓ | | ✓ |
| Qwen2.5 | ✓ | | ✓ | | ✓ |
| Gemma 2 | ✓ | | | ✓ | |
| DeepSeek-V3 | ✓ | | ✓ | | ✓ |
| BLOOM | | ✓ | | | |
| MPT | | ✓ | | | |

## Interface Design (FlashInfer-style)

All GQA Ops share a unified parameter interface. RoPE frequencies are computed inside the kernel from `rope_theta` + `rope_scale` (no external cos/sin tables needed). ALiBi slopes are derived from `num_heads` internally.

```python
class GQAFwdOp(Op):
    def __init__(
        self,
        batch, heads, heads_kv, seq_len, dim,
        is_causal=True,
        pos_encoding_mode="NONE",   # "NONE" / "ROPE_LLAMA" / "ALIBI"
        window_left=-1,             # -1 = unlimited
        logits_soft_cap=None,       # None = disabled; float = tanh soft cap
        rope_scale=1.0,             # RoPE interpolation scale
        rope_theta=1e4,             # RoPE base frequency
        dtype=torch.float16,
    )

class GQADecodeOp(Op):
    def __init__(
        self,
        batch, heads, heads_kv, seqlen_kv, dim,
        pos_encoding_mode="NONE",
        window_left=-1,
        logits_soft_cap=None,
        rope_scale=1.0,
        rope_theta=1e4,
        q_scale=None,               # FP8 calibration
        k_scale=None,
        v_scale=None,
        dtype=torch.float16,
    )
```

The Op dispatches to the correct kernel based on these parameters:

```python
def _select_kernel(self):
    fp8 = self.dtype in (torch.float8_e4m3fn, torch.float8_e5m2)
    parts = ["GQA"]
    if fp8:
        parts.append("FP8")
    if self.pos_encoding_mode == "ROPE_LLAMA":
        parts.append("RoPE")
    elif self.pos_encoding_mode == "ALIBI":
        parts.append("ALiBi")
    if self.logits_soft_cap is not None:
        parts.append("LogitsClip")
    parts.append("Fwd")  # or "Decode", "DecodePaged", etc.
    base = "".join(parts)
    if is_hopper():
        return self.kernel_map[base + "WGMMAPipelinedKernel"]
    return self.kernel_map[base + "Kernel"]
```

## Naming Convention

Following the [PEP 8 naming analysis](https://github.com/superAngGao/agent-assisted-tileops/blob/brainstorm-naming-summary/brainstorm_naming/naming_convention_analysis.md): acronyms are fully capitalized (`GQA`, `FP8`, `ALiBi`, `WGMMA`), non-acronyms are PascalCase.

Feature segments are concatenated in fixed order: **FP8 → RoPE|ALiBi → LogitsClip → SlidingWindow → Varlen**

Examples:
- `GQARoPEFwdKernel`
- `GQAFP8RoPELogitsClipDecodePagedKernel`
- `GQARoPESlidingWindowVarlenFwdWGMMAPipelinedKernel`

## Split-KV Design

Split-KV is separated into independent kernels (not auto-dispatched inside a single kernel class). The Op selects split vs no-split based on `seqlen_kv` at runtime.

| Kernel type | Responsibility |
|-------------|---------------|
| No-split kernel (e.g. `GQADecodeKernel`) | Short sequences, single-pass computation |
| SplitKV kernel (e.g. `GQADecodeSplitKVKernel`) | Long sequences, outputs partial_o + partial_lse |
| Combine kernel (e.g. `GQADecodeCombineKernel`) | Merges partial results via online softmax |

Combine kernels are feature-agnostic (shared across all feature combinations).

### Combine Kernels (3 total, in `flash_attn/combine.py` and `flash_decode/combine.py`)

| Kernel | Location |
|--------|----------|
| `GQAFwdCombineKernel` | `flash_attn/` |
| `GQADecodeCombineKernel` | `flash_decode/` |
| `GQADecodePagedCombineKernel` | `flash_decode/` |

### Prefill Fwd — `flash_attn/` (48 kernels)

Each feature combination × {no-split, splitKV} × {base, Hopper WGMMA} = 4 kernels per combination.

| # | Combination | File |
|---|-------------|------|
| 1 | Base (existing) | `fwd.py` |
| 2 | RoPE | `rope_fwd.py` |
| 3 | ALiBi | `alibi_fwd.py` |
| 4 | RoPE + LogitsClip | `rope_logits_clip_fwd.py` |
| 5 | FP8 + RoPE | `fp8_rope_fwd.py` |
| 6 | FP8 + RoPE + LogitsClip | `fp8_rope_logits_clip_fwd.py` |
| 7 | SlidingWindow (move from deepseek_nsa/) | `sliding_window_fwd.py` |
| 8 | RoPE + SlidingWindow | `rope_sliding_window_fwd.py` |
| 9 | FP8 + RoPE + SlidingWindow | `fp8_rope_sliding_window_fwd.py` |
| 10 | SlidingWindow Varlen (move from deepseek_nsa/) | `sliding_window_varlen_fwd.py` |
| 11 | RoPE + SlidingWindow Varlen | `rope_sliding_window_varlen_fwd.py` |
| 12 | FP8 + RoPE + SlidingWindow Varlen | `fp8_rope_sliding_window_varlen_fwd.py` |

12 combinations × 4 = 48 kernels.

### Prefill Bwd — `flash_attn/` (8 kernels)

No split-KV, no FP8, no SlidingWindow for backward. Each × {base, Hopper} = 2.

| # | Combination | File |
|---|-------------|------|
| 1 | Base (existing) | `bwd.py` |
| 2 | RoPE | `rope_bwd.py` |
| 3 | ALiBi | `alibi_bwd.py` |
| 4 | RoPE + LogitsClip | `rope_logits_clip_bwd.py` |

4 combinations × 2 = 8 kernels.

### Decode Contiguous — `flash_decode/` (16 kernels)

Each combination × {no-split, splitKV} = 2.

| # | Combination | File |
|---|-------------|------|
| 1 | Base (existing) | `gqa_decode.py` |
| 2 | RoPE | `gqa_rope_decode.py` |
| 3 | ALiBi | `gqa_alibi_decode.py` |
| 4 | RoPE + LogitsClip | `gqa_rope_logits_clip_decode.py` |
| 5 | RoPE + SlidingWindow | `gqa_rope_sliding_window_decode.py` |
| 6 | FP8 + RoPE | `gqa_fp8_rope_decode.py` |
| 7 | FP8 + RoPE + LogitsClip | `gqa_fp8_rope_logits_clip_decode.py` |
| 8 | FP8 + RoPE + SlidingWindow | `gqa_fp8_rope_sliding_window_decode.py` |

8 combinations × 2 = 16 kernels.

### Decode Paged — `flash_decode/` (16 kernels)

Mirrors Decode Contiguous with `Paged` suffix.

| # | Combination | File |
|---|-------------|------|
| 1 | Base (existing) | `gqa_decode_paged.py` |
| 2 | RoPE | `gqa_rope_decode_paged.py` |
| 3 | ALiBi | `gqa_alibi_decode_paged.py` |
| 4 | RoPE + LogitsClip | `gqa_rope_logits_clip_decode_paged.py` |
| 5 | RoPE + SlidingWindow | `gqa_rope_sliding_window_decode_paged.py` |
| 6 | FP8 + RoPE | `gqa_fp8_rope_decode_paged.py` |
| 7 | FP8 + RoPE + LogitsClip | `gqa_fp8_rope_logits_clip_decode_paged.py` |
| 8 | FP8 + RoPE + SlidingWindow | `gqa_fp8_rope_sliding_window_decode_paged.py` |

8 combinations × 2 = 16 kernels.

## Summary

| Category | Count |
|----------|-------|
| Combine kernels | 3 |
| Prefill Fwd | 48 |
| Prefill Bwd | 8 |
| Decode Contiguous | 16 |
| Decode Paged | 16 |
| **Total** | **91** |

Existing: ~14. **New: ~77.**

## Implementation Priority

| Priority | Scope | Covers |
|----------|-------|--------|
| **P0** | RoPE (all phases) + SplitKV refactor | LLaMA, Mistral, Qwen, DeepSeek, Gemma |
| **P1** | RoPE + SlidingWindow (all phases) | Mistral, Qwen2.5, DeepSeek-V3 |
| **P2** | RoPE + LogitsClip | Gemma 2 |
| **P3** | FP8 + RoPE (decode + prefill fwd) | FP8 inference for all RoPE models |
| **P4** | ALiBi | BLOOM, MPT |

## File Reorganization

Sliding Window kernels should be moved from `deepseek_nsa/` to `flash_attn/` since sliding window is a general GQA feature, not DeepSeek-specific. `deepseek_nsa/` retains only NSA-specific kernels (compress, topk, etc.).
